### DISCUSSION TITLE
[面試實戰D4] 想問 (1)關於 Session (2) Browsers 端的問題

### DISCUSSION THREAD

**Discussion by BoisonChang** [Lidemy/forum#31] [前端相關問題]
想發問前幾天的問題（這邊可以討論問題嗎）

1. 我一直覺得 Browsers 端，直觀感覺聯想很像就是在形容單純的 Safari Chrome IE Firefox
但它代表的意思不只如此對不對，可以再怎麼解釋呀？因為我也覺得 Browsers 和 Clients 在不同地方的定義好像有點混淆...

2. Session 有點搞不懂他是一個用來認證 Clients 的機制設計 / 還是本身就是描述儲存在 Server 的認證資料，怎樣解釋最好？

**Comment by aszx87410** [comment]
1. browser 就是你講的那個意思沒錯，就是瀏覽器沒錯，所以我猜你真正想問的是 client 端？client 端是相對於 server 端來說的，所以像是瀏覽器或是你的手機都可以稱作是 client 端
2. 都是，這點其實在 session cookie 三部曲的第一篇就有提到了

> Session 是什麼？就是一種讓 Request 變成 stateful 的機制。以小明的例子來說，Session 就是一種讓客人之間能互相關聯起來的機制。在故事裡面我們用了紙條跟手機裡的資訊來比喻，有多種方式可以達成 Session。

> 而這個狀態儲存的地方在口語上也會被稱之為「Session」，例如說：「幫我把 user id 存在 session 裡」，或者是「登出記得把 session 清空」之類的，所以在實際用法中，我認為 session 之所以不好理解是因為太多地方用到同一個詞，但卻是在指涉不同的東西（可是又很類似）。跟 API 有點像，太多地方都用到這個詞了。

就像 API 你也很難講清楚是什麼一樣，session 我覺得也很難講清楚，因為有很多意思而且有時候其實在描述類似但不同的事物

也可以參考這個：https://www.youtube.com/watch?v=DceuW3gxbNg

**Reply by BoisonChang** [reply]
https://blog.huli.tw/2017/03/12/csrf-introduction/

老師你在這篇，前面提到使用者，最後才提到 Browsers，所以我有點不確定你指的使用者是 users 還是 clients

而若使用者是指 users ，那.......

Browsers 等於 Clients 嗎？還是在某些情境下相同，

**Reply by aszx87410** [reply]
使用者是指 users 沒錯，不會是指 client，client 會被翻譯成客戶端，使用者自己的防禦指的就是 user 的防禦

browsers 不等於 clients，就像我前面講的，clients 的定義更廣泛，但在講前後端溝通，尤其是跟瀏覽器相關的溝通的時候，browser 就是一個 client，你可以用 server client 當關鍵字去查一下相關資料

**Reply by Lauviah0622** [reply]
Client 跟 Server 是主從式架構的兩個角色，網際網路採用了這個架構，所以才會稱呼平常向 server 拿資料的的角色就是 client 。使用者（user）會透過 browser 向 server 拿資料，然後 browser 收到之後再呈現這些資料給 user。

browser 應該可以說是一種應用程式？可以幫我們發 request 執行收回來的 response 之類的東西。也不只有 Safari Chrome IE Firefox 是 browser 而已，只要符合上面的功能應該可以稱作是 browser 吧，有一些 browser 好像也沒有介面，可以提供類似爬蟲之類來使用資料或者是其他用途。

自己的理解大概是這樣XD

**Reply by BoisonChang** [reply]
感謝，完全通透的理解了！
